### DISCUSSION TITLE
Deep deleting frames + labels to make shorter .slp file

### DISCUSSION THREAD

**Discussion by catubc** [talmolab/sleap#1076] [Ideas]
Hello
We're working on a project using sleap bottom_up with ID to automatically detect identity swaps in predictions. The idea is:
- we use sleap prediction to find identity swaps (i.e. two neighbouring frames where the features of an animal don't move a lot, but the identity swaps)
- make a list of these pairs of frames
- make a hybrid video of just these pairs of frames
- make a hybrid .slp file of preditions on these pairs of frames.

Here's what the video looks like:
https://drive.google.com/file/d/10dlNPkS5G6t4Lf2Fm39r5bBj3n-keHTw/view?usp=sharing

The next step is to have a human go over these hybrid .slp files and fix these errors. We think this will be 10-100x faster than labeling single frames randomly and it appears to be a principled way to find places where the sleap NNs are getting confused.

However, we're a bit stuck on how to do a deep delete on an .slp file where we:
- delete the video frame - (don't know how to do this)
- delete the predicted instance  (this I think we got from: https://github.com/talmolab/sleap/discussions/1044)
- overwrite filename -> hybrid video file name (don't know how to do this)
- and all the other things that need to be done to get a good .slp file that is only 50-100 frames (instead of 30000 frames as in our videos).

If this works, we're hoping to develop it into a feature that perhaps sleap engineers could code up.
Thanks so much,
@catubc 

So far we have this code and it runs, but it only deletes predicted instances, the video is still just as big as the original video:

```python

fname = 'file.slp'

#
labels = sleap.load_file(fname)
n_videos = len(labels.videos)

# 
matched_frame = 1000    #Keep only a single frame in the .slp file
animal_name = "animal0"  # the only instance to keep in frame 1000

#
for video_idx in range(n_videos):
    
    # Change this select the video of interest from the labels.video list
    video = labels.videos[video_idx]
    labeled_frames_from_video = labels.get(video)
        
    # Loop through all labeled frames (from a specific video)
    n_frames = 0
    ctr1=0
    for lf in tqdm(labeled_frames_from_video):
        
        # some LOGIC/CONDITIONAL - DOESN"T MATTER
        if ctr1==matched_frame:

            #
            n_inst = np.arange(len(lf.predicted_instances))
            
            #
            for k, inst in zip(n_inst, lf.predicted_instances):
                
                #
                name_predicted = lf.predicted_instances[k].track.name
                
                # delete the non-matched instances:
                if name_predicted != animal_name:
                    lf.predicted_instances.remove(inst)
               
        # IF NOT <- DEEP DELETE THE ENTIRE FRAME + VIDEO FRAME + PREDICTION + ALL INSTANCES
        else:
            labels.remove_frame(lf)   # <------------- NEED CORRECT CODE HERE            
        ctr1+=1


    # RENAME the video file
    labels.videos[video_idx] = 'file_short.slp'

    # Save the modified slp
    new_filename = fname[:-4]+"_deleted.slp"  # Use a different path from the current slp to be safe
    labels.save_file(labels, new_filename)
print ("DONE...")
```

So this code should in principle output an .slp file with a single prediction at frame 1000 and be redirected to a new filename called 'file_short.slp' - which we'll have made separately (<- unless sleap can also do this for us).

**Comment by talmo** [comment]
Hi @catubc,

Sorry, can you clarify what you mean by "delete the video frame"? Like you want to edit the MP4 to not have a particular frame?

The `labels.remove_frame()` will take out the `sleap.LabeledFrame` from the `sleap.Labels`, but the `sleap.LabeledFrame` is just a reference to a `sleap.Video` object and a frame index, not the frame image itself.

Also, in your code where you're doing `labels.videos[video_idx] = "file_short.slp"`, this is going to be an error. The `labels.videos` list expects to contain a list of `sleap.Video` objects, not a string pointing a .slp file. It's what you get back from `sleap.load_video("video.mp4")` for example. A .slp file is NOT a video, it's a saved `sleap.Labels` object.

If you need to edit the actual video pixels, you'll need some logic for this, but I think you already have that based on the rendered video with the background blacked out.

Let me know if this helps or if we can clarify further!

Talmo

**Comment by roomrys** [comment]
Hi @catubc,

Great idea and initiative! Proofreading is something that could definitely be improved upon.

The way I see this integrating with SLEAP:
---
1. Create a metric to measure likelihood of switches
2. Create a config file for a new Proofreading dialog (see Labeling Suggestions config below)
https://github.com/talmolab/sleap/blob/eac2e2bc2e7a98f61dd8efc90540433da69e846f/sleap/config/suggestions.yaml#L1-L193
3. Create a Model for Proofreading dialog (similar to Labeling Suggestions) which displays frames where switches are likely to occur.
https://github.com/talmolab/sleap/blob/eac2e2bc2e7a98f61dd8efc90540433da69e846f/sleap/gui/dataviews.py#L466-L548
4. Create the Proofreading dialog in the app
https://github.com/talmolab/sleap/blob/eac2e2bc2e7a98f61dd8efc90540433da69e846f/sleap/gui/app.py#L1069-L1150
5. Add commands that link to buttons in the Proofreading dialog
https://github.com/talmolab/sleap/blob/eac2e2bc2e7a98f61dd8efc90540433da69e846f/sleap/gui/commands.py#L556-L558
6. Add correct update topics to `on_data_update`
https://github.com/talmolab/sleap/blob/eac2e2bc2e7a98f61dd8efc90540433da69e846f/sleap/gui/app.py#L1287-L1360

For your own implementation:
---
Seeing as you only want to retain a select few frames (`select_frames` << `len(video)`), I would suggest first gathering the frames to keep, then looping through those frames instead of looping through the entire video.

```python
# Find indices of labeled frames to keep, note these are the indices of the `LabeledFrame`s in the labels.labeled_frames list
lf_indices = [0, 2, 7, 101]  # Replace with list of labeled frame indices (NOT the same as video indices)
tracks_to_keep = ["animal0", "animal0", "animal1", "animal0"]  # List of track names to keep (same size as `lf_indices`)

# Create a new `Labels` object containing only the matched `LabeledFrame`s
labels_keep = labels.extract(lf_indices, copy=True)

# Loop through matched `LabeledFrame`s
for lf in labels_keep.labeled_frames:
    for inst_idx, inst in enumerate(lf.predicted_instances):
        # Your code here

# Save new slp
new_slp_path = "path/new/slp"
sleap.Labels.save_file(labels_keep, new_slp_path)
```

This won't create a new video, but you can more easily jump to the frames where the switches are more likely to occur (use Alt + L/R Arrow Key).

**Comment by catubc** [comment]
Ok, thank you both for the advice.  We're getting pretty close now, only 1 step left: deleting specific tracks in each predicted frame that we keep.  

1.  We're going to simplify the task and not bother to reshape the .slp file. So (on a predicted .slp file) we'll 
A) delete all predicted frames that we're not interested in (99%)
B) in the remaining frames we'll delete the animal tracks we're also not interested in.

2. @roomrys code already solves (A), but B we can't seem to figure out how.  I'm confused about  the in-place deletion, maybe you  could helps us there?
 
I attach the code snippet + a screen grab of what we want.  As you can see the .slp file has only 15-20 predicted frames (which is good), we just want to also delete very specific tracks within the remaining frames.

Thanks so much,

```
#
labels = sleap.load_file(fname)
n_videos = len(labels.videos)

#
n_all=0
for video_idx in range(n_videos):
    
    # Change this select the video of interest from the labels.video list
    video = labels.videos[video_idx]
    labeled_frames_from_video = labels.get(video)
    
    # Find indices of labeled frames to keep, note these are the indices of the `LabeledFrame`s in the labels.labeled_frames list
    lf_indices = search_frames[:,0] #[0, 2, 7, 101]  # Replace with list of labeled frame indices (NOT the same as video indices)
    
    # Create a new `Labels` object containing only the matched `LabeledFrame`s
    labels_keep = labels.extract(lf_indices, copy=True)

    # Loop through matched `LabeledFrame`s
    ctr=0
    for lf, track_to_keep in zip(labels_keep.labeled_frames, tracks_to_keep):
        
        #
        for inst_idx, inst in enumerate(lf.predicted_instances):
            
            # get the predicted name
            name_predicted = lf.predicted_instances[inst_idx].track.name

            #  see if it matches the track to keep name 
            if name_predicted!=track_to_keep:
                lf.predicted_instances.remove(inst)                                                   #<----- dosn't work
                del labels_keep.labeled_frames[ctr].predicted_instances[inst_idx]   #<-------------these functions don'tseem to work
                labels_keep.labeled_frames[ctr].predicted_instances.remove(inst)    #  i.e. the saved file has all the tracks still
        ctr+=1
        
# Save new slp
new_slp_path = fname[:-4]+"_deleted.slp"
labels_keep.save_file(labels_keep, new_slp_path)                          #<----- these files still have all the tracks in each predicted instance
sleap.Labels.save_file(labels_keep, new_slp_path[:-4]+"_2.slp")    #<----- same
```
![sleap_delete](https://user-images.githubusercontent.com/4267452/206662808-e1797d42-b499-40f6-b255-e722af313b17.png)

**Reply by roomrys** [reply]
Hi @catubc,

### Removing `Instance`s (from the `LabeledFrame` object)
You need to remove the instance from the `lf.instances` list. The `lf.predicted_instances` is just a property based off `lf.instances` (not editable directly):
https://github.com/talmolab/sleap/blob/ebd2e1ec8f062efaf2588884bb46e92a8030ddcd/sleap/instance.py#L1431-L1434

Try `lf.instances.remove(inst)`. This will remove the `Instance` from `labels_keep`, but it will not remove the `Track` from the project. 

### Removing `Track`s (from the `Labels` object) and references to the `Track` (from the `Instance` object)
If you want to remove the track from the project, use `labels_keep.tracks.remove(track_to_remove)` where `track_to_remove` references the actual `Track` object, not just the track name. You may want to check that no instances in your `labels_keep` project reference the `track_to_remove` before actually removing it. This is the `Labels.remove_track` method used in SLEAP (callable through `labels_keep.remove_track`):
https://github.com/talmolab/sleap/blob/ebd2e1ec8f062efaf2588884bb46e92a8030ddcd/sleap/io/dataset.py#L1235-L1240
Note that the above function first removes the references to that track `inst.track = None` and then removes the track `self.tracks.remove(track)`.

Let me know if you need anything else.

Thanks,
Liezl

**Comment by catubc** [comment]
And @roomrys , thanks for the description of the steps for integrating this into Sleap. If it works, we're happy to help out implementing this.
